src/net/sourceforge/kolmafia/request/RelayRequest.java

```diff
@@ -3500,6 +3514,156 @@ public boolean pollForCompletion(String type) {
     return true;
   }
 
+  private void jsonError(String message) {
+    this.statusLine = "HTTP/1.1 400 Bad Request";
+    this.contentType = "application/json";
+    this.responseCode = 400;
+    this.responseText = JSON.toJSONString(Map.of("error", message));
+  }
+
+  private String underscoreName(String name) {
+    return name.replaceAll("[A-Z]", "_$0").toLowerCase();
+  }
+
+  private Value transformApiArgument(Object thing) {
+    if (thing instanceof JSONObject obj) {
+      var objectTypeObject = obj.get("objectType");
+      Type dataType;
+      if (objectTypeObject instanceof String objectType
+          && ((dataType = DataTypes.enumeratedTypes.find(objectType.toLowerCase())) != null)) {
+        var identifierStringObject = obj.get("identifierString");
+        var identifierNumberObject = obj.get("identifierNumber");
+        if (identifierStringObject instanceof String identifierString) {
+          return dataType.parseValue(identifierString, true);
+        } else if (identifierNumberObject instanceof Integer identifierNumber) {
+          return dataType.makeValue(identifierNumber, true);
+        }
+      }
+
+      return JSONValueConverter.fromJSON(obj);
+    } else return JSONValueConverter.fromJSON(thing);
+  }
+
+  /**
+   * Handle JSON API request. Format is e.g. { properties: ["kingLiberated"], functions: [{ name:
+   * "myTunrcount", args: [] }}. Enumerated types (Item, Familiar, etc.) can be passed in with
+   * either of the following placeholder formats: { "objectType": "item", "identifierString":
+   * "seal-clubbing club" } { "objectType": "item", "identifierNumber": 22 } Enumerated types will
+   * be returned in the same format. Details for an enumerated type instance can be request with the
+   * special "identity" function, which takes one placeholder argument.
+   *
+   * @param request Request body
+   */
+  private void handleJsonApi(String request) {
+    this.contentType = "application/json";
+    JSONObject json;
+    try {
+      json = JSON.parseObject(request);
+    } catch (JSONException e) {
+      jsonError("Invalid JSON object in request.");
+      return;
+    }
+
+    var result = new JSONObject();
+
+    var properties = json.get("properties");
+    if (properties instanceof JSONArray propertiesArray) {
+      var nonString = propertiesArray.stream().filter(name -> !(name instanceof String)).toList();
+      if (!nonString.isEmpty()) {
+        jsonError("Invalid property names " + JSON.toJSONString(nonString));
+        return;
+      }
+      result.put(
+          "properties",
+          ((JSONArray) properties)
+              .stream()
+                  .map(name -> name instanceof String ? Preferences.getString((String) name) : null)
```

Comment:
the previous check confirmed the list contains no non-strings, so this instanceof check is unnecessary